File: pi4/c/examples/2.7.0_Ultrasonic/ultrasonic.c
/*
 This example works with the Mosiwi Basic learning board.

 Learning target:
	1. ultrasonic
	2. pluse

 Web: http://mosiwi.com
 Wiki: https://mosiwi-wiki.readthedocs.io
 Designer: jalen
 Date：2022-4-3
 */
#include <stdio.h>
#include <wiringPi.h>

#define Trig 23
#define Echo 24

float ultr(void){
	float distance;
	unsigned int t;
	int cnt = 0;
	// The Trig pin sends a 10uS high level trigger pulse signal.
	digitalWrite(Trig, HIGH);
	delayMicroseconds(10);
	digitalWrite(Trig, LOW);

	while(!digitalRead(Echo)){
		delayMicroseconds(10);
		cnt ++;
		if(cnt > 5100){
			cnt = 0;
			break;
		}
	}
	t = micros();
	while(digitalRead(Echo)){
		delayMicroseconds(10);
		cnt ++;
		if(cnt > 5100){
			cnt = 0;
			break;
		}
	}
	t = micros() - t;

	// Convert microseconds to seconds: rxTime/1000000.0  --->  1s = 1000000us
	// The speed of ultrasound: 340m/s = 34000cm/s,
	// Because the data measured by ultrasound has two distances back and forth, the data has to be divided by 2.
	// Convert the pulse signal to a distance in centimeters.
	distance = ((t/1000000.0) * 34000.0) / 2;
	return distance;
}

// the setup function runs once when you press reset or power the board
int main(void){
	wiringPiSetup();

	// Example Initialize the I/O port
	pinMode(Trig, OUTPUT);
	pinMode(Echo, INPUT);
	digitalWrite(Trig, LOW);

	float dis = 0;
	while(1){
		dis = ultr();

		// Print distance
		printf("distance: %.2f cm \n", dis);
		delay(500);
	}
}

File: pico/python/ultrasonic.py
import time
from machine import Pin, PWM

echo = Pin(9, Pin.IN)       # create input pin on GPIO9, S1 pin of the expansion board
trig = Pin(8, Pin.OUT)      # create output pin on GPIO8, S2 pin of the expansion board
trig.value(0)

def Measuring_distance():
    c = 0
    # Send a 10us trigger signal to the ultrasonic module.
    trig.value(1)
    time.sleep_us(20)
    trig.value(0)
    
    # Capture the pulse signal of the echo pin of the ultrasonic module.
    while(echo.value() == 0 and c <= 1000):
        time.sleep_us(20)
        c = c + 1
    if c >= 1000:
        return 0.0
    c = 0
    while(echo.value() == 0 and c <= 800 ):
        time.sleep_us(5)
        c = c + 1
    if c >= 800:
        return 0.0
    else:
        return 1.7*c  # Unit is centimeter
    
    
while True:
    distance = Measuring_distance()
    print("Ultrasonic distance measured: %0.1f CM" %distance)
    time.sleep_ms(1000)
    
    
    